akka-http-core/src/test/scala/akka/http/impl/util/CollectionStage.scala

```diff
@@ -0,0 +1,127 @@
+ * It will return all the collected elements and a flag whether the stream was completed or not.
+ *
+ * When this is added to a fused flow it can be used to check what output and completion a given input
+ * will bring. Because the side-channel will go through the mailbox of the ActorGraphInterpreter the
+ * hope is that the fused stream will "run to exhaustion", i.e. it will run as far as it can, so that
+ * when the side-channel event is processed, no other (internal) events are still to be processed. This
+ * assumption does not hold if the source stream has asynchronous boundaries or components.
+ *
+ * TODO: allow multiple rescheduling of AsyncCallback to make it more likely that all internal queues
+ * have been processed.
+ */
+@InternalApi
+private[akka] class CollectorStage[T] extends GraphStageWithMaterializedValue[SinkShape[T], Collect[T]] {
+  val in = Inlet[T]("CollectorStage.in")
+  def shape: SinkShape[T] = SinkShape(in)
+
+  /** State of the callback infrastructure */
+  sealed trait State
+  case object Uninitialized extends State
+  case class Scheduled(callback: Promise[(Seq[T], Boolean)]) extends State
+  case object Initialized extends State
+  case class Completed(elements: Try[(Seq[T], Boolean)]) extends State
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Collect[T]) = {
+    val logic = new GraphStageLogic(shape) with Collect[T] with InHandler {
+      var collectedElements = new VectorBuilder[T]()
+      val state = new AtomicReference[State](Uninitialized)
+
+      val collectNowCallback = getAsyncCallback[Promise[(Seq[T], Boolean)]] { promise ⇒
+        require(state.get().isInstanceOf[Scheduled])
+        promise.complete(Success(collectedElements.result() → false))
+        completeStage()
+      }
+
+      setHandler(in, this)
+      override def preStart(): Unit = {
+        pull(in)
+
+        @tailrec def handleState(): Unit =
+          state.get() match {
+            case Uninitialized ⇒
+              if (!state.compareAndSet(Uninitialized, Initialized)) handleState()
+            case early @ Scheduled(callback) ⇒
+              // make sure to go through asynchandler -> actor interpreter mailbox
+              Future(collectNowCallback.invoke(callback))(materializer.executionContext)
+            case s ⇒ throw new IllegalStateException(s"Unexpected state $s")
+          }
+        handleState()
+      }
+      def onPush(): Unit = {
+        val data = grab(in)
+        pull(in)
+        collectedElements += data
+      }
+
+      def collectAndCompleteNow(): Future[(Seq[T], Boolean)] = {
+        val p = Promise[(Seq[T], Boolean)]()
+
+        @tailrec def handleState(): Unit =
+          state.get() match {
+            case Initialized ⇒
+              if (!state.compareAndSet(Initialized, Scheduled(p))) handleState()
+              else collectNowCallback.invoke(p)
```

Comment:
Had to move the callback invocation here so that it cannot overtake the state change (or be even invoked multiple times).